File: backend/app/schemas/protocol.py
from typing import Literal

from pydantic import BaseModel


class ProtocolParameterResponse(BaseModel):
    name: str
    label: str
    data_type: Literal["integer"]
    default: int
    minimum: int
    maximum: int
    unit: str | None = None
    description: str


class ProtocolResponse(BaseModel):
    id: str
    name: str
    description: str
    parameters: list[ProtocolParameterResponse]


class ProtocolListResponse(BaseModel):
    items: list[ProtocolResponse]

File: frontend/types/protocol.ts
export type ProtocolParameterDataType = "integer";

export type ProtocolParameter = {
  name: string;
  label: string;
  data_type: ProtocolParameterDataType;
  default: number;
  minimum: number;
  maximum: number;
  unit: string | null;
  description: string;
};

export type ProtocolDefinition = {
  id: string;
  name: string;
  description: string;
  parameters: ProtocolParameter[];
};

export type ProtocolListResponse = {
  items: ProtocolDefinition[];
};